src/Raven.Server/Documents/Subscriptions/SubscriptionConnectionBase.cs

```diff
@@ -0,0 +1,490 @@
+
+                if (translation.TryGet(nameof(Client.Documents.Subscriptions.SubscriptionState.SubscriptionId), out long id) == false)
+                    throw new SubscriptionClosedException("Could not figure out the Subscription Task ID for subscription named: " + _options.SubscriptionName);
+
+                SubscriptionId = id;
+            }
+        }
+
+        internal async Task WriteJsonAsync(DynamicJsonValue value)
+        {
+            int writtenBytes;
+            using (TcpConnection.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            await using (var writer = new AsyncBlittableJsonTextWriter(context, TcpConnection.Stream))
+            {
+                context.Write(writer, value);
+                writtenBytes = writer.Position;
+            }
+
+            await TcpConnection.Stream.FlushAsync();
+            TcpConnection.RegisterBytesSent(writtenBytes);
+        }
+
+        internal async Task SendHeartBeat(string reason)
+        {
+            try
+            {
+                await TcpConnection.Stream.WriteAsync(Heartbeat, 0, Heartbeat.Length, CancellationTokenSource.Token);
+                await TcpConnection.Stream.FlushAsync();
+
+                if (_logger.IsInfoEnabled)
+                {
+                    _logger.Info($"Subscription {Options.SubscriptionName} is sending a Heartbeat message to the client. Reason: {reason}");
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new SubscriptionClosedException($"Cannot contact client anymore, closing subscription ({Options?.SubscriptionName})", ex);
+            }
+
+            TcpConnection.RegisterBytesSent(Heartbeat.Length);
+        }
+
+        internal async Task<Task<SubscriptionConnectionClientMessage>> WaitForClientAck(Task<SubscriptionConnectionClientMessage> replyFromClientTask, bool sharded)
+        {
+            SubscriptionConnectionClientMessage clientReply;
+            while (true)
+            {
+                var result = await Task.WhenAny(replyFromClientTask,
+                    TimeoutManager.WaitFor(TimeSpan.FromMilliseconds(5000), CancellationTokenSource.Token)).ConfigureAwait(false);
+                CancellationTokenSource.Token.ThrowIfCancellationRequested();
+                if (result == replyFromClientTask)
+                {
+                    clientReply = await replyFromClientTask;
+                    if (clientReply.Type == SubscriptionConnectionClientMessage.MessageType.DisposedNotification)
+                    {
+                        CancellationTokenSource.Cancel();
+                        break;
+                    }
+
+                    replyFromClientTask = sharded ? null : GetReplyFromClientAsync();
+                    break;
+                }
+
+                await SendHeartBeat("Waiting for client ACK");
+                if (sharded == false)
+                {
+                    await SendNoopAck();
```

Comment:
should be overridden in sharded and be a no-op impl

Reply:
done